File: Tourism-BE/src/main/java/com/gialai/tourism/models/entities/Post.java
package com.gialai.tourism.models.entities;

import com.gialai.tourism.common.base.BaseEntity;
import com.gialai.tourism.enums.PostStatus;
import com.gialai.tourism.enums.SourceType;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Table(name = "posts")
@Getter @Setter
@NoArgsConstructor @AllArgsConstructor @Builder
public class Post extends BaseEntity {

    @Column(nullable = false, length = 255)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    @Column(length = 500)
    private String summary;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private SourceType sourceType;

    @Column(length = 255)
    private String sourceName;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PostStatus status;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    private Account author;

    @Column(nullable = false)
    private long viewCount;

    @Column(nullable = false)
    private long likeCount;

    @Column(nullable = false)
    private long favoriteCount;

    @Column(nullable = false)
    private double averageRating;

    @Column(nullable = false)
    private int ratingCount;

    @ElementCollection
    @CollectionTable(name = "post_images", joinColumns = @JoinColumn(name = "post_id"))
    @Column(name = "image_url")
    private List<String> images;

    @ManyToMany
    @JoinTable(name = "post_tags",
            joinColumns = @JoinColumn(name = "post_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id"))
    @Builder.Default
    private Set<Tag> tags = new HashSet<>();

    @OneToOne(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    private Location location;

    private LocalDateTime approvedAt;
    private LocalDateTime rejectedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "approved_by")
    private Account approvedBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "rejected_by")
    private Account rejectedBy;

    @Column(length = 500)
    private String rejectionReason;
}

File: Tourism-FE/src/types/post.ts
export type PostStatus = 'PENDING' | 'APPROVED' | 'REJECTED' | 'DELETED';
export type SourceType = 'AUTHOR' | 'EXTERNAL';
export type Tag =
  | 'LOCATION'
  | 'CULTURE'
  | 'HISTORY'
  | 'FESTIVAL'
  | 'FOOD'
  | 'ACCOMMODATION'
  | 'TRANSPORT';

export interface Post {
  id: string;
  title: string;
  content: string;
  summary?: string;
  thumbnail?: string;
  images: string[];
  tags: Tag[];
  sourceType: SourceType;
  sourceName?: string | null;
  authorUsername: string;
  authorFullName?: string;
  status: PostStatus;
  viewCount: number;
  likeCount: number;
  favoriteCount: number;
  averageRating: number;
  ratingCount: number;
  createdAt: string;
  updatedAt: string;
  approvedAt?: string;
  rejectedAt?: string;
  rejectedReason?: string;
}

export interface CreatePostPayload {
  title: string;
  content: string;
  summary?: string;
  tags: Tag[];
  images?: string[];
  sourceType: SourceType;
  sourceName?: string | null;
}

export type UpdatePostPayload = Partial<CreatePostPayload>;

export interface PostListResponse {
  content: Post[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
}
